keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -2026,6 +1947,29 @@ type PodResourceClaimStatus struct {
 }
 ```
 
+##### ResourceQuota
+
+The core ResourceQuota mechanism checks at admission time whether creating a
+resource in the apiserver would exceed per-namespace limits. This mechanism
+gets extended to support limiting the number of ResourceClaims based on what
+gets requested.
+
+A resource entry in ResourceQuota with the special
+`<device-class-name>.deviceclass.resource.k8s.io/devices` limits the number of
+ResourceClaim objects in a namespace such that the number of devices requested
+through those objects with that class does not exceed the limit. When admins
+define classes so that they select specific device types, this quota mechanism
+can be used to limit access to those device types.
+
+A single request may cause the allocation of multiple devices. For exact
+counts, the quota limit is based on the sum of those exact counts. For requests
+asking for "all" matching devices, the maximum number of allocated devices per
+claim is used as a worst-case upper bound.
+
+Requests asking for "admin access" contribute to the quota. In practice,
+namespaces where such access is allowed will typically not have quotas
+configured.
```

Comment:
Is this just because it is simpler to implement? I would think that since the resources are not actually "allocated" they shouldn't count against quota.

Reply:
Yes, for the sake of simplicity. It allowed removing one if check in the code. I had it initially and was asked to remove it.

Reply:
Still seems odd to me (because you essentially end up double counting the same resource) but OK.

Reply:
@johnbelamaric @thockin what was the rationale for having these semantics?

Reply:
https://github.com/kubernetes/kubernetes/pull/120611#discussion_r1688013919

> Still seems odd to me (because you essentially end up double counting the same resource)

Counting is by namespace. Users in their namespace have a quota and so might admins (if enabled in their namespace, which probably won't happen). That it's counting the same device doesn't make a difference, it's about "how much do you get access to" in both cases.

Reply:
I agree that this is unlikely to matter in practice, so easier / simpler is better

Reply:
+1